### PR TITLE
refactor(datasource/adoptium-java): Strict null checks

### PR DESCRIPTION
**Pull request by zharinov**
## Changes:

- Extract page fetching method
- Refactor `do-while` to `while`
- Make it ready for strict null checks

## Context:

Blocked by #13416

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/datasource/adoptium-java/index.spec.ts [hunk 1]
```diff
@@ -10,13 +10,6 @@ const jre = loadFixture('jre.json');
 function getPath(page: number, imageType = 'jdk'): string {
   return `/v3/info/release_versions?page_size=${pageSize}&image_type=${imageType}&project=jdk&release_type=ga&sort_method=DATE&sort_order=DESC&vendor=adoptium&page=${page}`;
 }
-function* range(start: number, end: number): Generator<number, number, number> {
```
Comment (viceice) [inline lib/datasource/adoptium-java/index.spec.ts:13]: What's the issue about this function?

File: lib/datasource/adoptium-java/index.ts [hunk 2]
```diff
@@ -25,6 +25,29 @@ export class AdoptiumJavaDatasource extends Datasource {
 
   override readonly caching = true;
 
+  private async getPageReleases(
+    url: string,
+    page: number
+  ): Promise<Release[] | undefined> {
```
Comment (viceice) [inline lib/datasource/adoptium-java/index.ts:31]: We should use null instead of undefined

File: lib/datasource/adoptium-java/index.spec.ts (hunk 1 above)
Comment (zharinov) [reply lib/datasource/adoptium-java/index.spec.ts:13]: Just wasn't sure if my fix would be right

File: lib/datasource/adoptium-java/index.spec.ts [hunk 3]
```diff
@@ -10,7 +10,11 @@ const jre = loadFixture('jre.json');
...
+
+function* range(
+  start: number,
+  end: number
+): Generator<number, number, number> | undefined {
```
Comment (viceice) [inline lib/datasource/adoptium-java/index.spec.ts:17]: Under what conditions it will return undefined? 🤔

File: tsconfig.strict.json [hunk 4]
```diff
@@ -90,6 +91,7 @@
   ],
   "exclude": [
     "lib/constants/platform.spec.ts",
+    "lib/datasource/adoptium-java/**/*.spec.ts",
```
Comment (viceice) [inline tsconfig.strict.json:94]: Too hard?

File: lib/datasource/adoptium-java/index.ts [hunk 5]
```diff
@@ -25,6 +25,31 @@ export class AdoptiumJavaDatasource extends Datasource {
...
+          version: semver,
+        })) ?? null
+      );
+    } catch (err) {
+      if (
+        page !== 0 &&
+        err instanceof HttpError &&
+        err.response?.statusCode === 404
+      ) {
+        return null;
```
Comment (viceice) [inline lib/datasource/adoptium-java/index.ts:46]:
```suggestion
        // No more pages
        return null;
```
Add the lost comment

File: lib/datasource/adoptium-java/index.ts [hunk 6]
```diff
@@ -46,24 +70,20 @@ export class AdoptiumJavaDatasource extends Datasource {
...
-          await this.http.getJson<AdoptiumJavaResponse>(`${url}&page=${page}`)
-        ).body;
-        result.releases.push(
-          ...resp.versions.map(({ semver }) => ({ version: semver }))
-        );
+      let page = 0;
+      let releases = await this.getPageReleases(url, page);
+      while (releases) {
+        result.releases.push(...releases);
+        if (releases.length !== pageSize) {
```
Comment (viceice) [inline lib/datasource/adoptium-java/index.ts:78]:
```suggestion
        if (releases.length !== pageSize || page >= 50) {
```
Lost condition.

File: tsconfig.strict.json (hunk 4 above)
Comment (zharinov) [reply tsconfig.strict.json:94]: Usually I try to include test files as well, but in this case it transitively depends on the entire `lib/versioning/` directory.

Comment (viceice) [reply tsconfig.strict.json:94]: Ok, so we should fix versionings first.

File: lib/datasource/adoptium-java/index.spec.ts (hunk 3 above)
Comment (zharinov) [reply lib/datasource/adoptium-java/index.spec.ts:17]: Didn't think `void` can be used here. Changed to `void`.

File: tsconfig.strict.json (hunk 4 above)
Comment (zharinov) [reply tsconfig.strict.json:94]: Also it depends on `lib/datasource/index.ts`, so it seems to be a circular situation. My reasoning behind listing `*.spec.ts` files in the `exclude` section is that their null errors are mostly fixed locally, i.e. once all non-test files are refactored, spec files from `exclude` section should work automatically.

File: lib/datasource/adoptium-java/index.spec.ts (hunk 3 above)
Comment (viceice) [inline lib/datasource/adoptium-java/index.spec.ts:17]:
```suggestion
): IterableIterator<number> {
```
Maybe this works?

https://stackoverflow.com/a/68945873

Comment (viceice) [inline lib/datasource/adoptium-java/index.spec.ts:17]: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-6.html#stricter-generators

I think this is the right definition:

```suggestion
): Generator<number, void, voud> {
```

**Review by rarkins — APPROVED** [review]

**Pull request merged**